Cache Fake and Dummy constructors

blairconrad opened the pull request:
Fixes #1359.

blairconrad commented:
I can certainly squash some commits later if you like, but left the Fake ones exploded because they are a little messy.

812cc3a5abd1e6ad570cc6eba4a0945cd299b781 is purely refactoring, mostly done because I found that treating the parameterless constructor the same way as a constructor used due to constructor arguments being passed was confusing.
Of course, it does support the later changes.

thomaslevesque commented on src/FakeItEasy/Creation/FakeObjectCreator.cs:
```diff
@@ -115,9 +118,10 @@ private ProxyGeneratorResult TryCreateFakeWithDummyArgumentsForConstructor(Type
             var consideredConstructors = new List<ResolvedConstructor>();
-            foreach (var parameterTypes in GetUsableParamterTypeListsInOrder(typeOfFake))
+
+            if (this.parameterTypesCache.TryGetValue(typeOfFake, out Type[] cachedParameterTypes))
```
Again, use `GetOrAdd` if possible? If it's too awkward, forget it

blairconrad commented on src/FakeItEasy/Core/DefaultExceptionThrower.cs:
```diff
@@ -50,21 +50,26 @@ public void ThrowFailedToGenerateProxyWithResolvedConstructors(Type typeOfFake,
+                    if (constructor.Arguments == null)
+                    {
+                        message
+                            .AppendIndented("    ", "No constructor arguments failed:");
```
I don't love it either. I'd prefer "Paramterless constructor failed", or similar, or remove the special case and just indicate that it's the one with the signature `()`.

blairconrad replied:
However, I'm preserving the existing message. Happy to change, but would rather do in another issue, for visibility.

thomaslevesque replied:
> remove the special case and just indicate that it's the one with the signature `()`.

I like that!

> However, I'm preserving the existing message.

Ah, I didn't realize. Sure, it can be done in a separate PR.

blairconrad replied:
#1373

blairconrad commented on src/FakeItEasy/Creation/ResolvedConstructor.cs:
```diff
@@ -6,7 +6,7 @@ internal class ResolvedConstructor
         public bool WasSuccessfullyResolved
         {
-            get { return !this.Arguments.Any(x => !x.WasResolved); }
+            get { return this.Arguments == null || !this.Arguments.Any(x => !x.WasResolved); }
```
Yeah, I hear you. And I saw the warning as well. I've been bitten by taking a `!Any` to `All` type of transform before and was too lazy to properly evaluate this time.
Although with the caveat that we will always have at least one argument in the list, I think you're right. Will do.

blairconrad commented on src/FakeItEasy/Creation/DummyValueResolver.cs:
```diff
@@ -218,21 +220,20 @@ public override bool TryCreateDummyValue(DummyCreationSession session, Type type
                 }
 
-                foreach (var constructor in GetConstructorsInOrder(typeOfDummy))
+                if (this.cachedConstructors.TryGetValue(typeOfDummy, out ConstructorInfo cachedConstructor))
```
This was my thought. We'd need an action that created (located) the constructor and then returned the Dummy as a side effect. It grates a little.
I know `GetOrAdd` prevents concurrent resolution, but at least there's no harm in concurrent resolution aside from a small amount of wasted effort.

All that being said, I'll try to see if I can find a palatable solution.

blairconrad replied:
Hmm... I know in Java the concurrent hashtable isn't reentrant. I'll have to make sure we don't have the same problem here.

thomaslevesque replied:
I don't know if `ConcurrentDictionary` supports reentrancy... Anyway, if it turns out to be to complex, just leave it as is.

blairconrad commented on tests/FakeItEasy.Specs/CreationSpecs.cs:
```diff
@@ -94,6 +95,69 @@ public interface IInterfaceWithSimilarMethods
+                .See(() => new ClassWhosePreferredConstructorsThrow());
+
+            "And the class has a two-parameter constructor that throws"
+                .See(() => new ClassWhosePreferredConstructorsThrow(A.Dummy<IDisposable>(), A.Dummy<string>()));
```
Happy to change. But there are two wrinkles:

thomaslevesque replied:
Which version of the IDE are you using? I think 7.1 is supported in VS2017 15.3.

blairconrad replied:
I'm not at it right now, but I'm _sure_ it's newer. And for some reason, it's stopped complaining about the `default`. I don't know what its deal is.

blairconrad commented:
Thanks for the comments, @thomaslevesque. I think I've addressed everything except the `GetOrAdd` request. I have doubts that it'll work out, but am open to trying it over the next coupla days.
(Oh, while we're on about it, `TryResolveDummyValueWithAllAvailableStrategies` suffers from the same problem.)

blairconrad commented:
> I think I've addressed everything except the GetOrAdd request.

Oops. And the wording around [parameterless constructors](https://github.com/FakeItEasy/FakeItEasy/pull/1372#discussion_r196890730). If it suits, I'll spin off another issue for that.

thomaslevesque commented on src/FakeItEasy/Creation/FakeObjectCreator.cs:
```diff
@@ -60,32 +60,34 @@ private static IEnumerable<ResolvedConstructor> ResolveConstructors(Type typeOfF
         }
 
-        private static IEnumerable<ConstructorInfo> GetUsableConstructorsInOrder(Type type)
+        private static IEnumerable<Type[]> GetUsableParamterTypeListsInOrder(Type type)
```
I think you forgot to fix this one ;)

blairconrad commented:
Squashed the fixup!s and pushed (I also corrected one more "paramter" while I was there). Could still squash more, if you like. Compare with if you're curious.

I've not yet given up on `GetOrAdd`. I just wanted to tidy up first.

thomaslevesque commented:
> I've not yet given up on GetOrAdd. I just wanted to tidy up first.

OK, I'll wait for your go before I merge then

blairconrad commented:
Okay, I pushed a solution that uses `GetOrAdd`. It's a little rough, and could be cleaned up a little (using a list as a Maybe is not the best, and so on), but honestly, ![I'm not a huge fan](https://imgs.xkcd.com/comics/turbine.png). 

I think it's convoluted (even more convoluted than the original `TryCreateFakeWithDummyArgumentsForConstructor`!). There's a possibility that it'll save a little work when concurrent threads try to resolve the same Fake or Dummy type at the same time, but it's unlikely to save more than one attempt, and I think most times it'll end up saving nothing.

Then again, maybe there's a better way than my approach, but any solution is going to have to distinguish between "cached constructor/set of types/resolver that I just now figured out, so I have already materialized a fake/dummy" and "brand new constructor/set of types/resolver that I have to invoke to get a fake/dummy", and that makes the solutions, well, icky.

blairconrad commented:
Loving DontMergeMeYet, BTW.

blairconrad commented:
Whoops! I'd just run `build -s build spec` at home before pushing. How silly. Looks like I broke a test. I can try to figure out in the morning, unless we hate the approach enough already to just drop it (I think I do, but happy to see what you think).

thomaslevesque commented:
I'm not a big fan of this solution either. Let's just drop it, there's not much to be gained. Thanks for trying anyway!

blairconrad commented:
Thanks, @thomaslevesque. Unpushed.

thomaslevesque commented:
Thanks @blairconrad!

blairconrad commented:
Thanks for the merge. And review!
